pkg/instancetype/revision/store_test.go

```diff
@@ -0,0 +1,561 @@
+
+			createdCR, err := virtClient.AppsV1().ControllerRevisions(vm.Namespace).Get(
+				context.Background(), vm.Spec.Instancetype.RevisionName, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(createdCR).To(Equal(instancetypeControllerRevision))
+		})
+
+		It("store fails when instancetype does not exist", func() {
+			vm.Spec.Instancetype.Name = nonExistingResourceName
+			err := storeHandler.Store(vm)
+			Expect(err).To(HaveOccurred())
+			Expect(errors.IsNotFound(err)).To(BeTrue())
+		})
+
+		It("store returns a nil revision when RevisionName already populated", func() {
+			instancetypeControllerRevision, err := revision.CreateControllerRevision(vm, fakeInstancetype)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(
+				context.Background(), instancetypeControllerRevision, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+				Name:         fakeInstancetype.Name,
+				RevisionName: instancetypeControllerRevision.Name,
+				Kind:         apiinstancetype.SingularResourceName,
+			}
+
+			Expect(storeHandler.Store(vm)).To(Succeed())
+			Expect(vm.Spec.Instancetype.RevisionName).To(Equal(instancetypeControllerRevision.Name))
+		})
+
+		It("store ControllerRevision succeeds if a revision exists with expected data", func() {
+			instancetypeControllerRevision, err := revision.CreateControllerRevision(vm, fakeInstancetype)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(
+				context.Background(), instancetypeControllerRevision, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(storeHandler.Store(vm)).To(Succeed())
+			Expect(vm.Spec.Instancetype.RevisionName).To(Equal(instancetypeControllerRevision.Name))
+		})
+
+		It("store ControllerRevision fails if a revision exists with unexpected data", func() {
+			unexpectedInstancetype := fakeInstancetype.DeepCopy()
+			unexpectedInstancetype.Spec.CPU.Guest = 15
+
+			instancetypeControllerRevision, err := revision.CreateControllerRevision(vm, unexpectedInstancetype)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(
+				context.Background(), instancetypeControllerRevision, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(storeHandler.Store(vm)).To(MatchError(ContainSubstring("found existing ControllerRevision with unexpected data")))
+		})
+
+		It("store ControllerRevision fails if instancetype conflicts with vm", func() {
+			vm.Spec.Template.Spec.Domain.CPU = &virtv1.CPU{
+				Cores: 1,
+			}
+			Expect(storeHandler.Store(vm)).To(MatchError(conflict.Conflicts{conflict.New("spec", "template", "spec", "domain", "cpu", "cores")}))
+		})
+	})
+	Context("using global ClusterPreference", func() {
```

Comment:
IMO the ClusterPreference Contexts/tests cases belong into the Preference Context above? (`store preference`).

Reply:
ACK sorry missed this, reworked things under the `store` `Contexts`.